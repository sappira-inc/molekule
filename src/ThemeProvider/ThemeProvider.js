import React from 'react';
import { ThemeProvider as SCThemeProvider } from 'styled-components';
import merge from 'lodash/merge';
import defaultTheme from '../theme';

const ThemeProvider = props => {
  const theme = merge(defaultTheme(props.theme), props.theme);

  return <SCThemeProvider theme={theme}>{props.children}</SCThemeProvider>;
};

export default ThemeProvider;
