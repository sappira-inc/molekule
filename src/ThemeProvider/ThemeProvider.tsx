import React, { FC } from 'react';
import { ThemeProvider as SCThemeProvider, DefaultTheme } from 'styled-components';
import merge from 'lodash/merge';
import { createTheme } from '../createTheme';

interface ThemeProviderProps {
  theme?: Partial<DefaultTheme>;
}

export const ThemeProvider: FC<ThemeProviderProps> = props => {
  const theme = merge(createTheme(props.theme), props.theme);

  return <SCThemeProvider theme={theme}>{props.children}</SCThemeProvider>;
};
