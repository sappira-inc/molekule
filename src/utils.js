import { get, kebabCase } from 'lodash';
import styled from 'styled-components';

export const themeGet = (theme, lookup, fallback) => get(theme, lookup, fallback);

export const getComponentVariant = (theme, componentName, variant) => {
  const config = themeGet(
    theme,
    `components.${componentName}.variants.${variant}`,
    theme.variants[componentName][variant]
  );
  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentStyle = componentName => ({ theme }) => themeGet(theme, `components.${componentName}.style`);

export const getComponentClassName = ({ className, theme: { classPrefix }, variant }, name) =>
  `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

export const createComponent = ({ name, tag = 'div', as, style, props: defaultProps = () => ({}) }) => {
  const component = as ? styled(as) : styled[tag];

  return component.attrs(props => ({
    ...defaultProps(props),
    ...props,
    className: getComponentClassName(props, kebabCase(name)),
  }))`
    ${style}
    ${getComponentStyle(name)}
    ${({ styles = {} }) => styles[name] || {}}
  `;
};
