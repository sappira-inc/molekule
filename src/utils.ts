import kebabCase from 'lodash/kebabCase';
import get from 'lodash/get';
import styled, { CSSProp } from 'styled-components';

export const themeGet = (lookup: any, fallback?: any) => ({ theme }: any = {}) => get(theme, lookup, fallback);

export const getComponentVariant = (theme: any, componentName: string, variant: string) => {
  const config = themeGet(`variants.${componentName}.${variant}`)({
    theme,
  });

  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentSize = (theme: any, componentName: string, size: string) =>
  themeGet(`sizes.${componentName}.${size}`, {})({ theme });

export const getComponentStyle = (componentName: string) => themeGet(`styles.${componentName}`, {});

export const getVariantStyles = (componentName: string, variant: string) =>
  themeGet(`variants.${componentName}.${variant}.style`, {});

interface GetComponentClassNameProps {
  className: string;
  theme: {
    classPrefix: string;
  };
  variant: string;
}

const getComponentClassName = (
  { className, theme: { classPrefix }, variant }: GetComponentClassNameProps,
  name: string
) => `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

interface CreateComponentProps<T> {
  name: string;
  tag?: keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>;
  as?: React.ComponentType<any>;
  style?: ((props: any) => CSSProp) | CSSProp;
  props?: (props: T) => any;
}

export const createComponent = <
  T extends object,
  O extends keyof JSX.IntrinsicElements | React.ComponentType<any> = 'div'
>({
  name,
  tag = 'div',
  as,
  style,
  props: getBaseProps = () => ({}),
}: CreateComponentProps<T>) => {
  const component = styled<O>((as || tag) as any);

  return component.attrs<T>((props: any) => {
    const baseProps = getBaseProps(props);
    const finalProps = {
      ...baseProps,
      ...props,
      className: baseProps.className || props.className,
    };

    return {
      ...finalProps,
      className: getComponentClassName(finalProps, kebabCase(name)),
    };
  })<T>`
  ${style}
  ${getComponentStyle(name)}
  ${(p: any) => getVariantStyles(name, p.variant)}
  ${({ styles = {} }) => styles[name] || {}}
`;
};

// eslint-disable-next-line
const focusableFilter = (node: any) =>
  node.tabIndex >= 0 && !node.hasAttribute('disabled') ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP;

type FilterFn = (node?: any) => any;

const createTreeWalker = (root: any, currentNode: any, filterFn: FilterFn = () => NodeFilter.FILTER_ACCEPT) => {
  const treeWalker = document.createTreeWalker(
    root,
    NodeFilter.SHOW_ELEMENT,
    {
      acceptNode: filterFn,
    },
    false
  );
  treeWalker.currentNode = currentNode;
  return treeWalker;
};

export const findPreviousFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.previousNode() || treeWalker.currentNode === root) {
    return treeWalker.lastChild();
  }

  return treeWalker.currentNode;
};

export const findNextFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.nextNode() || treeWalker.currentNode === root) {
    treeWalker.currentNode = root;
    return treeWalker.firstChild();
  }

  return treeWalker.currentNode;
};

export const getNextCursorPosition = (cursorPos: any, oldValue: any, newValue: any) => {
  const cursorDiff = newValue.length - oldValue.length;
  let nextPosition = cursorPos;

  if (cursorDiff > 1 || cursorDiff < -1) {
    nextPosition += cursorDiff < 0 ? cursorDiff + 1 : cursorDiff - 1;
  }

  return Math.max(nextPosition, 0);
};
