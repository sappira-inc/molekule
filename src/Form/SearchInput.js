import React, { useRef, useState } from 'react';
import { css } from 'styled-components';
import { createComponent } from '../utils';
import { createEasyInput } from './EasyInput';
import Input from './Input';

const Search = createComponent({
  name: 'SearchInput',
  as: Input,
  style: ({ theme, size }) => css`
    color: ${theme.colors.grayDark};
    background-color: ${theme.colors.grayLight};
    height: ${theme.heights[size]}px;
    caret-color: ${theme.colors.greyDark};
    border-color: ${theme.colors.grayLightest};
    font-size: ${theme.fontSizes[size]}px;
    padding-left: 28px;
    padding-right: 28px;

    &:hover,
    &:active,
    &:focus {
      border-color: ${theme.colors.grayLight};
    }
  `,
});

function SearchInput({ forwardedRef, value: propValue, onChange, ...inputProps }) {
  const [currentValue, setValue] = useState(propValue);
  const inputRef = forwardedRef || useRef();

  const handleChange = (name, newValue, event) => {
    setValue(newValue);

    if (onChange) {
      onChange(name, newValue, event);
    }
  };

  return (
    <Search
      forwardedRef={inputRef}
      icon="magnify"
      onChange={handleChange}
      value={currentValue}
      isClearable
      {...inputProps}
    />
  );
}

export default createEasyInput(SearchInput);
