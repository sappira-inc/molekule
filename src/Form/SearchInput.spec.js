import React from 'react';
import { renderWithTheme } from '../../test/utils';
import SearchInput from './SearchInput';
import ThemeProvider from '../ThemeProvider';

describe('<SearchInput />', () => {
  const renderInput = props => {
    const utils = renderWithTheme(<SearchInput placeholder="Search" label="Search" {...props} />);
    return {
      ...utils,
      input: utils.getByPlaceholderText('Search'),
    };
  };

  test('snapshot', () => {
    const { asFragment } = renderInput();
    expect(asFragment()).toMatchSnapshot();
  });

  test('initally sets the value', () => {
    const value = 'Search item';
    const { input } = renderInput({ value });

    expect(input.value).toEqual('Search item');
  });

  test('updates internally when value prop changes', () => {
    const { input, rerender } = renderInput({ value: 'Searching' });
    rerender(
      <ThemeProvider>
        <SearchInput placeholder="Search" />
      </ThemeProvider>
    );
    expect(input.value).toEqual('Searching');
  });
});
