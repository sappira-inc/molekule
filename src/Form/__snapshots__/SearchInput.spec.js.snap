// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchInput /> snapshot 1`] = `
<DocumentFragment>
  .c1 {
  position: relative;
  -webkit-transition: 175ms;
  transition: 175ms;
}

.c1 + .c0 {
  margin-top: 1rem;
}

.c2 {
  display: block;
  -webkit-transition: 250ms;
  transition: 250ms;
  font-weight: 500;
  margin: 0 0 4px 4px;
  font-size: 12px;
}

.c3 {
  position: relative;
}

.c6 {
  border: 1px solid #DEE0E4;
  height: 32px;
  display: block;
  outline: none;
  width: 100%;
  padding: 8px 24px 8px 8px;
  border-radius: 8px;
  -webkit-transition: 250ms all;
  transition: 250ms all;
  -webkit-appearance: none;
  font-family: inherit;
  font-size: 12px;
  padding-left: 28px;
}

.c6:hover,
.c6:focus,
.c6:active {
  border-color: #8E97A7;
}

.c6::-webkit-input-placeholder {
  color: #8E97A7;
}

.c6::-moz-placeholder {
  color: #8E97A7;
}

.c6:-ms-input-placeholder {
  color: #8E97A7;
}

.c6::placeholder {
  color: #8E97A7;
}

.c6[disabled] {
  opacity: 0.65;
}

.c4 {
  color: #43526D;
  font-size: 16px;
  position: absolute;
  left: 8px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c5 {
  font-size: 14px;
  color: #43526D;
  background-color: #DEE0E4;
  caret-color: #43526D;
  border-color: #F1F4F6;
}

.c5:hover,
.c5:active,
.c5:focus {
  border-color: #DEE0E4;
}

<div
    class="c0 re-field c1"
  >
    <label
      class="re-label c2"
    >
      Search
    </label>
    <div
      class="re-input-container c3"
    >
      <i
        class="mdi mdi-magnify re-icon re-left-icon c4"
        color="greyDarkest"
        data-testid="left-icon"
        name="magnify"
        size="16"
      />
      <input
        class="re-search-input c5 sc-iwsKbI re-search-input c5 re-input c6"
        placeholder="Search"
        rows="3"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
